### PR TITLE
migration to sqlalchemy 2.0

### PR DESCRIPTION
**Pull request by farahats9**
made the required changes to pass the tests for sqlalchemy 2.0
python version had to be pumped to 3.7 instead of 3.6
removed sqlalchemy2-stubs since it is not compatible or required.

More testing in a real project scenario is welcome.

### CODE REVIEW

**Comment by farahats9** [comment]
I fixed some of the `mypy` linting errors but I am not good at these things so maybe someone can fix the rest.

File: sqlmodel/sql/expression.py [hunk 1]
```diff
@@ -26,34 +24,15 @@
 
 # Workaround Generics incompatibility in Python 3.6
 # Ref: https://github.com/python/typing/issues/449#issuecomment-316061322
```
Comment (sbor23) [inline sqlmodel/sql/expression.py]:
```suggestion

```

**Comment by honglei** [comment]
I used this version in my project with env: `Python3.11.2/sqlalchemy2.0.7/Win10/scheme="postgresql+asyncpg"`, it works all fine.

File: docs/contributing.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ If you already cloned the repository and you know that you need to deep dive in
 
 ### Python
 
-SQLModel supports Python 3.6 and above, but for development you should have at least **Python 3.7**.
+SQLModel supports Python 3.7 and above, but for development you should have at least **Python 3.7**.
```
Comment (sbor23) [inline docs/contributing.md]: Second part of that sentence doesn't make sense anymore.

File: docs/tutorial/index.md [hunk 3]
```diff
@@ -85,7 +85,6 @@ You might want to try with the specific versions, for example with:
 * `python3.9`
 * `python3.8`
 * `python3.7`
-* `python3.6`
```
Comment (sbor23) [inline docs/tutorial/index.md:88]: Might wanna add `python3.11` for consistency.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -54,7 +54,7 @@ jobs:
         if: steps.cache.outputs.cache-hit != 'true'
         run: python -m poetry install
       - name: Lint
-        if: ${{ matrix.python-version != '3.6.15' }}
+        if: ${{ matrix.python-version != '3.7' }}
```
Comment (sbor23) [inline .github/workflows/test.yml]: That line can probably go away? Looks like a workaround for 3.6

**Comment by gabrielmbmb** [comment]
@farahats9 someone created a [PR on your fork](https://github.com/farahats9/sqlmodel/pull/1) for fixing the mypy issues

**Comment by starkana** [comment]
This seems good to go just needs merging

**Comment by 8thgencore** [comment]
We are very much waiting for this change

**Comment by antont** [comment]
> We are very much waiting for this change

Why do you wait, instead of taking it to use? I guess feedback from users would be helpful info also for the maintainer to know when to merge.

**Comment by linpan** [comment]
🚀 👀👀

**Comment by k-s-t-i** [comment]
Would like to see this merged for asyncpg support. Asyncpg seems to greatly increase performance based on most simple benchmarks.

**Comment by farahats9** [comment]
> Would like to see this merged for asyncpg support. Asyncpg seems to greatly increase performance based on most simple benchmarks.

You can currently use Asyncpg with sqlalchemy v1.4 with no issue in most cases. you will need to watch out for some problems like lazy loading and cascading deletes on the orm side.

**Comment by synodriver** [comment]
> Great! I would love to try this branch in my production environment.

Update: This patch can work with asyncmy in a real project. Will try asyncpg later.

File: pyproject.toml [hunk 5]
```diff
@@ -30,10 +31,9 @@ classifiers = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.6.1"
-SQLAlchemy = ">=1.4.17,<=1.4.41"
+python = "^3.7"
+SQLAlchemy = ">=2.0.0,<=2.0.5.post1"
```
Comment (synodriver) [inline pyproject.toml]: Can we remove ```<=2.0.5.post1```? I tested it with the latest version and it works fine.

**Comment by synodriver** [comment]
> > Great! I would love to try this branch in my production environment.
> 
> Update: This patch can work with asyncmy in a real project. Will try asyncpg later.

Update: asyncpg works too. Tested with a real project on both windows and linux. Notice: to make relationship work, you may use ```Relationship(back_populates="xxx", sa_relationship_kwargs={"lazy": 'selectin'})``` where ```sa_relationship_kwargs``` is important.

File: pyproject.toml (hunk 5 above)
Comment (UnoYakshi) [reply pyproject.toml]: It's SQLAlchemy 2.0.11 already. Should we really limit on 2.0.5?

File: sqlmodel/engine/result.py [hunk 6]
```diff
@@ -1,29 +1,29 @@
...
-    def partitions(self, size: Optional[int] = None) -> Iterator[List[_T]]:
+    def partitions(self, size: Optional[int] = None) -> Iterator[Sequence[_T]]:
         return super().partitions(size)
 
-    def fetchall(self) -> List[_T]:
+    def fetchall(self) -> Sequence[_T]:
         return super().fetchall()
 
-    def fetchmany(self, size: Optional[int] = None) -> List[_T]:
+    def fetchmany(self, size: Optional[int] = None) -> Sequence[_T]:
```
Comment (UnoYakshi) [inline sqlmodel/engine/result.py:19]: Apologize for my ignorance, but what's the purpose of changing `List` to `Sequence` here? Can it become a set or a dict somehow? Or is it for some other reason?

File: sqlmodel/main.py [hunk 7]
```diff
@@ -478,6 +478,7 @@ class SQLModel(BaseModel, metaclass=SQLModelMetaclass, registry=default_registry
     __sqlmodel_relationships__: ClassVar[Dict[str, RelationshipProperty]]  # type: ignore
     __name__: ClassVar[str]
     metadata: ClassVar[MetaData]
+    __allow_unmapped__ = True  # https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-20-step-six
```
Comment (UnoYakshi) [inline sqlmodel/main.py:482]: Thanks for putting the explanation reference!

File: sqlmodel/sql/expression.py [hunk 8]
```diff
@@ -24,36 +22,15 @@
...
-    # Cast them for editors to work correctly, from several tricks tried, this works
-    # for both VS Code and PyCharm
-    Select = cast("Select", _Py36Select)  # type: ignore
-    SelectOfScalar = cast("SelectOfScalar", _Py36SelectOfScalar)  # type: ignore
+# This is not comparable to sqlalchemy.sql.selectable.ScalarSelect, that has a different
+# purpose. This is the same as a normal SQLAlchemy Select class where there's only one
+# entity, so the result will be converted to a scalar by default. This way writing
+# for loops on the results will feel natural.
+class SelectOfScalar(_Select[_TSelect], Generic[_TSelect]):
+    inherit_cache = True
```
Comment (UnoYakshi) [inline sqlmodel/sql/expression.py]: Why not using a class dosctring for the note?

I'd also probably throw a reference link in the docstring on how to utilize it `SelectOfScalar`.

File: sqlmodel/sql/expression.py [hunk 9]
```diff
@@ -437,22 +387,21 @@ def select(  # type: ignore
...
-    **kw: Any,
 ) -> Select[Tuple[_TModel_0, _TModel_1, _TModel_2, _TModel_3]]:
     ...
 
 
 # Generated overloads end
 
 
-def select(*entities: Any, **kw: Any) -> Union[Select, SelectOfScalar]:  # type: ignore
+def select(*entities: Any) -> Union[Select, SelectOfScalar]:  # type: ignore
```
Comment (UnoYakshi) [inline sqlmodel/sql/expression.py:399]: Shouldn't the return annotation be something like `Union[Select[_T], SelectOfScalar[_T]`?

File: sqlmodel/sql/expression.py [hunk 10]
```diff
@@ -437,22 +387,21 @@ def select(  # type: ignore
...
     if len(entities) == 1:
-        return SelectOfScalar._create(*entities, **kw)  # type: ignore
-    return Select._create(*entities, **kw)  # type: ignore
+        return SelectOfScalar(*entities)  # type: ignore
+    return Select(*entities)  # type: ignore
```
Comment (UnoYakshi) [inline sqlmodel/sql/expression.py:402]: What's the issue with the `# type: ignore` here? Is it MyPy?
There's quite a lot of those statements all around. I'm not sure it's a healthy practice.

File: sqlmodel/__init__.py [hunk 11]
```diff
@@ -71,7 +70,6 @@
 from sqlalchemy.sql import outerjoin as outerjoin
 from sqlalchemy.sql import outparam as outparam
 from sqlalchemy.sql import over as over
-from sqlalchemy.sql import subquery as subquery
```
Comment (UnoYakshi) [inline sqlmodel/__init__.py:74]: Why?

File: docs_src/tutorial/relationship_attributes/back_populates/tutorial003.py [hunk 12]
```diff
@@ -3,6 +3,21 @@
...
+    secret_name: str
+    age: Optional[int] = Field(default=None, index=True)
+
+    team_id: Optional[int] = Field(default=None, foreign_key="team.id")
+    team: Optional["Team"] = Relationship(back_populates="heroes")
+
+    weapon_id: Optional[int] = Field(default=None, foreign_key="weapon.id")
+    weapon: Optional["Weapon"] = Relationship(back_populates="hero")
+
+    powers: List["Power"] = Relationship(back_populates="hero")
```
Comment (UnoYakshi) [inline docs_src/tutorial/relationship_attributes/back_populates/tutorial003.py:18]: Why putting `Hero` before `Weapon`? Is there any functional reasoning behind? Stringed annotations feel more noisy/dirty, IMO.

**Review by UnoYakshi — COMMENTED** [review]
Really would like to use it in my new project! Thank you for the PR!

File: sqlmodel/engine/result.py (hunk 6 above)
Comment (farahats9) [reply sqlmodel/engine/result.py:19]: The change from `List` to `Sequence` is because it inherits from sqlalchemy's `ScalarResult` and these are the types defined in the new version so they must be the same in sqlmodel

File: sqlmodel/sql/expression.py (hunk 8 above)
Comment (farahats9) [reply sqlmodel/sql/expression.py]: You are probably right, but this is copied from the original code as is 

https://github.com/tiangolo/sqlmodel/pull/563/files/050cf02a35b670eedeb814606d27869c3a723328#diff-976d5f3bdb62b7724d2e776bb785b98e13d96aac3a72d117b047467828756851L34-L39

File: sqlmodel/main.py (hunk 7 above)
Comment (farahats9) [reply sqlmodel/main.py:482]: np

File: sqlmodel/__init__.py (hunk 11 above)
Comment (farahats9) [reply sqlmodel/__init__.py:74]: no longer exists in `sqlalchemy.sql`

File: sqlmodel/sql/expression.py (hunk 10 above)
Comment (farahats9) [reply sqlmodel/sql/expression.py:402]: That's how it was before my code so I kept it as it is.

File: pyproject.toml (hunk 5 above)
Comment (farahats9) [reply pyproject.toml]: When I started this PR it was the latest version, I haven't tested it with more recent versions but will do soon.

File: docs_src/tutorial/relationship_attributes/back_populates/tutorial003.py (hunk 12 above)
Comment (farahats9) [reply docs_src/tutorial/relationship_attributes/back_populates/tutorial003.py:18]: I think this was due to circular import error when testing. Stringed annotations are a must if you intend to use sqlmodel or it will fail in complex structure models from experience.

File: sqlmodel/sql/expression.py (hunk 9 above)
Comment (farahats9) [reply sqlmodel/sql/expression.py:399]: Again keeping changes to original code minimal, I am only changing what's necessary to make sqlalchemy 2.0 work with sqlmodel. These type fixes that are not in the original code and will not affect v2.0 migration should be made in a separate PR IMO.

File: sqlmodel/__init__.py (hunk 11 above)
Comment (farahats9) [reply sqlmodel/__init__.py:74]: However, on second thoughts there is a new `Subquery` class added instead so I am adding it too.

**Comment by farahats9** [comment]
> > > Great! I would love to try this branch in my production environment.
> > 
> > 
> > Update: This patch can work with asyncmy in a real project. Will try asyncpg later.
> 
> Update: asyncpg works too. Tested with a real project on both windows and linux. Notice: to make relationship work, you may use `Relationship(back_populates="xxx", sa_relationship_kwargs={"lazy": 'selectin'})` where `sa_relationship_kwargs` is important.

Be careful! you don't want to do that all the time or you will have a lot of queries done when you don't need them. only add the `selectin` or `joined` flag when you know you will access this attribute for sure. otherwise use `noload` and add `options(orm.selectinload(...))`

**Comment by synodriver** [comment]
> > > > Great! I would love to try this branch in my production environment.
> > > 
> > > 
> > > Update: This patch can work with asyncmy in a real project. Will try asyncpg later.
> > 
> > 
> > Update: asyncpg works too. Tested with a real project on both windows and linux. Notice: to make relationship work, you may use `Relationship(back_populates="xxx", sa_relationship_kwargs={"lazy": 'selectin'})` where `sa_relationship_kwargs` is important.
> 
> @synodriver Be careful! you don't want to do that all the time or you will have a lot of queries done when you don't need them. only add the `selectin` or `joined` flag when you know you will access this attribute for sure. otherwise use `noload` and add `options(orm.selectinload(...))` when doing the select statement.

Actually the best way I think is to iterate over that attribute instead of loading them all. However, since I'm using async mod, it's imposible to do that. Maybe with raw sql?

**Comment by farahats9** [comment]
> Actually the best way I think is to iterate over that attribute instead of loading them all. However, since I'm using async mod, it's imposible to do that. Maybe with raw sql?

@synodriver Not sure what you mean by iterating over them I would need an example. 
In most situations you need to know which attributes you will use and load only these attributes. its usually fine to load additional attributes as long as they are in the same table, but if you are loading a relation then you need to add it in the query as an option as I explained (or in the model declaration if its always needed)

**Comment by synodriver** [comment]
> @synodriver Not sure what you mean by iterating over them I would need an example. In most situations you need to know which attributes you will use and load only these attributes. its usually fine to load additional attributes as long as they are in the same table, but if you are loading a relation then you need to add it in the query as an option as I explained (or in the model declaration if its always neede)

 Here is the models from the real project
```
from datetime import datetime
from typing import Optional, List

from sqlmodel import SQLModel, Field, PrimaryKeyConstraint, Relationship


class Game(SQLModel, table=True):
    __tablename__ = "game"
    game_id: Optional[int] = Field(None, primary_key=True)
    game_sale_time: Optional[datetime]
    game_translated_sale_time: Optional[datetime]
    game_collect_time: Optional[datetime]
    game_size: Optional[int]
    game_show_flag: Optional[int]
    game_delete_flag: Optional[int]
    game_index: Optional[int]
    game_name: str

    filesave: List["FileSave"] = Relationship(back_populates="game", sa_relationship_kwargs={"lazy": 'selectin'})
    photo: List["Photo"] = Relationship(back_populates="game", sa_relationship_kwargs={"lazy": 'selectin'})
    text: List["Text"] = Relationship(back_populates="game", sa_relationship_kwargs={"lazy": 'selectin'})


class FileSave(SQLModel, table=True):
    __tablename__ = "universal_file_save"
    universal_file_save_id: int = Field(None, primary_key=True)
    universal_file_save_a: Optional[int]
    universal_file_save_b: Optional[int] = Field(None, foreign_key="game.game_id")  # gameId
    universal_file_save_c: Optional[int]
    universal_file_save_url: Optional[str]
    universal_file_save_remark: Optional[int]
    universal_file_save_title: Optional[int]
    universal_file_save_rank: Optional[int]
    universal_file_save_show_flag: Optional[int]
    universal_file_save_delete_flag: Optional[int]
    game: Optional[Game] = Relationship(back_populates="filesave", sa_relationship_kwargs={"lazy": 'selectin'})


class Photo(SQLModel, table=True):
    __tablename__ = "universal_photo"
    universal_photo_id: int = Field(None, primary_key=True)
    universal_photo_a: Optional[int]
    universal_photo_b: Optional[int] = Field(None, foreign_key="game.game_id")  # gameId
    universal_photo_c: Optional[int]
    universal_photo_url: Optional[str]
    universal_photo_path: Optional[str]
    universal_photo_remark: Optional[str]
    universal_photo_time: Optional[datetime]
    universal_photo_rank: Optional[int]
    universal_photo_flag: Optional[int]
    universal_photo_show_flag: Optional[int]
    universal_photo_delete_flag: Optional[int]
    # relation ship
    game: Optional[Game] = Relationship(back_populates="photo", sa_relationship_kwargs={"lazy": 'selectin'})


class Text(SQLModel, table=True):
    __tablename__ = "universal_text"
    universal_text_id: int = Field(None, primary_key=True)
    universal_text_a: Optional[int]
    universal_text_b: Optional[int] = Field(None, foreign_key="game.game_id")  # gameId
    universal_text_c: Optional[int]
    universal_text_d: Optional[int]
    universal_text_e: Optional[int]
    universal_text_data: Optional[str]
    universal_text_show_flag: Optional[int]
    universal_text_delete_flag: Optional[int]
    game: Optional[Game] = Relationship(back_populates="text", sa_relationship_kwargs={"lazy": 'selectin'})
```

Those lists might be very big, so iterating over it may be  best choice. By the way, where to add ```options(orm.selectinload(...))```? I'm not very familiar with sqlalchemy...

**Comment by farahats9** [comment]
> where to add `options(orm.selectinload(...))`? I'm not very familiar with sqlalchemy...

@synodriver as an example you are trying to load `Game` object and the related `photo` atrribute without loading the others:

```python
from sqlalchemy import orm
from sqlmodel import select
from app.db.session import SessionLocal  # your own AsyncSession object


async def load_game(game_id):
    async with SessionLocal() as db:
        stmt = select(Game).select_from(Game).where(Game.game_id == game_id).options(orm.selectinload(Game.photo)).limit(1)
        game_obj = await db.scalar(stmt)
        # now you can access the game_obj.photo but not game_obj.text because it is not loaded.
```
hope this helps

File: sqlmodel/sql/expression.py (hunk 8 above)
Comment (UnoYakshi) [reply sqlmodel/sql/expression.py]: Roger. However, I think it's better to put the description in the class docstring. That way, it will be parsed/handled by whatever instrument usually does this (for example., your favourite IDE + some IntelliSense).

File: sqlmodel/sql/expression.py (hunk 10 above)
Comment (UnoYakshi) [reply sqlmodel/sql/expression.py:402]: Oh... That's a bit odd. To minimize the clutter, probably, we can turn off and on the code block with multiple `# type: ignore` lines?

File: sqlmodel/sql/expression.py (hunk 8 above)
Comment (UnoYakshi) [reply sqlmodel/sql/expression.py]: Until there is a valid reason from the original authors, of course.

**Comment by synodriver** [comment]
> @farahats9 Thanks for your help! I'd like to try the new option.

File: docs_src/tutorial/relationship_attributes/back_populates/tutorial003.py (hunk 12 above)
Comment (UnoYakshi) [reply docs_src/tutorial/relationship_attributes/back_populates/tutorial003.py:18]: Somewhat valid, yeah. I, too, experienced circular dependencies. Although, I had some workarounds (local imports mostly).

**Comment by mortendaehli** [comment]
Looking forward to this :) Appreciate the effort 🤩

**Comment by FilipeMarch** [comment]
I have one app I started last year with 21 tables,  I tried to use this PR but got this error:
```python
Traceback (most recent call last):
  File ".venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 221, in eval_expression
    annotation = eval(expression, base_globals, locals_)
  File "<string>", line 1, in <module>
NameError: name 'Article' is not defined

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/util.py", line 2329, in _extract_mapped_subtype
    annotated = de_stringify_annotation(
  File ".venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 156, in de_stringify_annotation
    annotation = eval_expression(
  File ".venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 223, in eval_expression
    raise NameError(
NameError: Could not de-stringify annotation 'Article'

During handling of the above exception, another exception occurred:


  File "./app/knowledge/models.py", line 202, in <module>
    class ArticleReviewerLink(SQLModel, table=True):
  File ".venv/lib/python3.10/site-packages/sqlmodel/main.py", line 369, in __init__
    DeclarativeMeta.__init__(cls, classname, bases, dict_used, **kw)
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_api.py", line 193, in __init__
    _as_declarative(reg, cls, dict_)
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 248, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 329, in setup_mapping
    return _ClassScanMapperConfig(
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 566, in __init__
    self._scan_attributes()
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 1011, in _scan_attributes
    collected_annotation = self._collect_annotation(
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 1285, in _collect_annotation
    extracted = _extract_mapped_subtype(
  File ".venv/lib/python3.10/site-packages/sqlalchemy/orm/util.py", line 2342, in _extract_mapped_subtype
    if raiseerr and "Mapped[" in raw_annotation:  # type: ignore
TypeError: argument of type 'ForwardRef' is not iterable
```
From this part of the code:
```python
class ArticleReviewerLink(SQLModel, table=True):
    __tablename__ = "article_reviewers"

    article_id: Optional[int] = Field(
        default=None, foreign_key="articles.id", primary_key=True
    )
    article: "Article" = Relationship(back_populates="reviewer_links")

    reviewer_id: Optional[int] = Field(
        default=None,
        foreign_key="reviewers.id",
        primary_key=True,
    )
    reviewer: "Reviewer" = Relationship(back_populates="article_links")

    reviewed: bool = Field(default=False)


class Article(SQLModel, table=True):
    __tablename__ = "articles"
    ...
```

I would love to be able to use sqlalchemy 2.0, but not sure how to solve this error. I would appreciate any help!

**Comment by antont** [comment]
> I have one app I started last year with 21 tables, I tried to use this PR but got this error:
> 
> ```python
> TypeError: argument of type 'ForwardRef' is not iterable

Just a note, that there's info in the SA release notes that sounds similar on the surface:

> Fixed issue where ORM Annotated Declarative would not resolve forward references correctly in all cases; in particular, when using from __future__ import annotations in combination with Pydantic dataclasses.

But the issue linked there does not look similar. Or?

20.0.13 notes in https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html link to -> 
https://github.com/sqlalchemy/sqlalchemy/discussions/9717

**Comment by podsvirov** [comment]
Hello @FilipeMarch! I've just change order of classes definition and solve your error:

> NameError: name 'Article' is not defined

Full example code (database.py):

```python
from typing import List, Optional
from sqlmodel import SQLModel, Field, Relationship, create_engine

class Article(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    reviewer_links: List["ArticleReviewerLink"] = Relationship(
        back_populates="article",
    )

class ArticleReviewerLink(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    article_id: Optional[int] = Field(default=None, foreign_key="article.id")
    article: Article = Relationship(back_populates="reviewer_links")

sqlite_file_name = "database.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"

engine = create_engine(sqlite_url, echo=True)

def create_db_and_tables():
    SQLModel.metadata.create_all(engine)

if __name__ == "__main__":
    create_db_and_tables()
````

File: pyproject.toml (hunk 5 above)
Comment (honglei) [reply pyproject.toml]: use `SQLAlchemy = ">=2.0.0,<=2.1"`  ?

**Comment by honglei** [comment]
Five months  have been passed since the release of SQLalchemy2.0, https://www.sqlalchemy.org/blog/2023/01/26/sqlalchemy-2.0.0-released/

**Comment by antont** [comment]
> Five months have been passed since the release of SQLalchemy2.0, https://www.sqlalchemy.org/blog/2023/01/26/sqlalchemy-2.0.0-released/

You can help by using this PR in your project(s) if you want, and reporting how well it works.

**Comment by 50Bytes-dev** [comment]
Been working with this PR for almost a month now, works great. Thank you!

**Comment by FilipeMarch** [comment]
> Hello @FilipeMarch! I've just change order of classes definition and solve your error:
> 
> > NameError: name 'Article' is not defined
> 
> Full example code (database.py):
> 
> ```python
> from typing import List, Optional
> from sqlmodel import SQLModel, Field, Relationship, create_engine
> 
> class Article(SQLModel, table=True):
>     id: Optional[int] = Field(default=None, primary_key=True)
>     reviewer_links: List["ArticleReviewerLink"] = Relationship(
>         back_populates="article",
>     )
> 
> class ArticleReviewerLink(SQLModel, table=True):
>     id: Optional[int] = Field(default=None, primary_key=True)
>     article_id: Optional[int] = Field(default=None, foreign_key="article.id")
>     article: Article = Relationship(back_populates="reviewer_links")
> 
> sqlite_file_name = "database.db"
> sqlite_url = f"sqlite:///{sqlite_file_name}"
> 
> engine = create_engine(sqlite_url, echo=True)
> 
> def create_db_and_tables():
>     SQLModel.metadata.create_all(engine)
> 
> if __name__ == "__main__":
>     create_db_and_tables()
> ```

Sorry but changing the order of classes definition does not solves the error for me. At this point, I don't know what to do, but this branch is useless for me. I already have done what you mentioned, but the error persists. It always says ```if raiseerr and "Mapped[" in raw_annotation:  # type: ignore
TypeError: argument of type 'ForwardRef' is not iterable```.  

My project has about 30 tables at this point, and it has been running for quite 1 year. My project does not run `as is` with this branch, and I already tried dozens of things (like switching the order of tables, removing Relationships, removing a bunch of tables) and I can't make it to work in any possible way. Errors keep coming non stop. I can deactivate the Relationships, move tables up or down, it does not matter, new errors come every time. I just feel wasting my time because obviously there is something wrong, but I can't stop the errors until I remove all tables from the code?

The error messages are not helpful at all. And they only happen in this specific PR.

I understand it may be working for a very simple case like you presented, but for me it is always showing these errors, no matter how much code I change:
```python
py", line 2479, in get_property
    return self._props[key]
KeyError: 'user'
```
or
```python
    annotation = eval(expression, base_globals, locals_)
  File "<string>", line 1, in <module>
NameError: name 'PostUserLike' is not defined
```
or
```python
.py", line 223, in eval_expression
    raise NameError(
NameError: Could not de-stringify annotation 'PostUserLike'
```
or
```python
", line 2342, in _extract_mapped_subtype
    if raiseerr and "Mapped[" in raw_annotation:  # type: ignore
TypeError: argument of type 'ForwardRef' is not iterable
```

My current API is running perfectly with `sqlmodel = "^0.0.8"`

If someone wants to investigate the issue with me, I am willing to screenshare and show the project structure and try to make it to work; my Discord is `filipemarch`

**Comment by farahats9** [comment]
@FilipeMarch  you are right there is an issue in `ForwardRef` handling and I am trying to investigate it when I have some free time.  The issue appears more when you have large project with many relations between tables so that's why some people report it working fine and some don't.

I feel this is a priority to fix over the mypy linting, I will do my best to figure it out but any help is appreciated.

**Comment by 50Bytes-dev** [comment]
I had an error `Mapped[" in raw_annotation` but I was able to solve it. Here is an example of my models, works fine. Maybe it will help you.

```python
# models/company.py

if TYPE_CHECKING:
    from models.location import Location
    from models.client import Client

class Company(BaseModel, CompanyBase, table=True):
    locations: List["Location"] = Relationship(back_populates="company")
    clients: List["Client"] = Relationship(back_populates="company")
```

```python
# models/location.py

from models.company import Company

class Location(BaseModel, OrderedList, LocationBase, table=True):
    company_id: int = Field(default=None, foreign_key="company.id")
    company: Company = Relationship(back_populates="locations")
```

```python
# models/client.py

from models.company import Company

class Client(ClientBase, BaseModel, table=True):
    company_id: int = Field(default=None, nullable=False, foreign_key="company.id")
    company: Company = Relationship(back_populates="clients")
```

**Comment by matanper** [comment]
Hey @farahats9,

I'm getting error when using sqlalchemy relationship:
```python
class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)

    powers: List["Power"] = Relationship(
        sa_relationship=relationship(back_populates='hero')
    )


class Power(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    hero_id: int = Field(foreign_key="hero.id")
    hero: "Hero" = Relationship(back_populates="powers")

```
```
NameError: name 'Power' is not defined
NameError: Could not de-stringify annotation 'Power'
```
any idea why?

**Comment by AntonDeMeester** [comment]
@matanper I think you the forward Relationship (with the Foreign Key) needs to be a proper class, and not a type string. Can you try this?

```python
class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)

    powers: List["Power"] = Relationship(
        sa_relationship=relationship(back_populates='hero')
    )


class Power(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    hero_id: int = Field(foreign_key="hero.id")
    hero: Hero = Relationship(back_populates="powers") # <== No quotes around Hero
```

**Comment by peterlandry** [comment]
I've looked at this PR a bit to try cleaning up the linting errors. I've reduced most of them, but `execute`, `get`, and `query` are a little complicated because of the overloads. One thing I noticed in the process though is that sqlalchemy 2.0 has enough typing info to break mypy in query code. In VS code, given a sqlalchemy ORM model, if I do:

```python
q = User.name == "spongebob"
```

`q` becomes an instance of `ColumnElement[bool]`, while if `User` is a SQLModel, `q` goes straight to a `bool` as far as mypy is concerned.

This shows up as an error, but is correct at runtime. At runtime, we see an `InstrumentedAttribute` for the attribute, but mypy thinks it's a string. I'm going to look further at that, and in the meantime submit a pull request here that fixes the majority of the linting errors.

**Comment by honglei** [comment]
Environment:
```  
pydantic version: 1.10.11
python version: 3.11.4 (main, Jun 27 2023, 15:03:25) [GCC 8.3.0]
platform: Linux-4.19.0-amd64-desktop-x86_64-with-glibc2.28
optional deps. installed: ['devtools', 'email-validator', 'typing-extensions']
sqlalchemy: 2.0.19
```

pytest errors:
```
================================================================================================================================================================================================================== short test summary info ==================================================================================================================================================================================================================
FAILED tests/test_tutorial/test_create_db_and_table/test_tutorial001.py::test_create_db_and_table - assert 'BEGIN' in "No module named 'docs_src'\n"
FAILED tests/test_tutorial/test_fastapi/test_delete/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...or'}}, ...}}}} == {'components'...or'}}, ...}}}}
FAILED tests/test_tutorial/test_fastapi/test_limit_and_offset/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...Read Hero'}}}} == {'components'...Read Hero'}}}}
FAILED tests/test_tutorial/test_fastapi/test_multiple_models/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...eate Hero'}}}} == {'components'...eate Hero'}}}}
FAILED tests/test_tutorial/test_fastapi/test_multiple_models/test_tutorial002.py::test_tutorial - AssertionError: assert {'components'...eate Hero'}}}} == {'components'...eate Hero'}}}}
FAILED tests/test_tutorial/test_fastapi/test_read_one/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...Read Hero'}}}} == {'components'...Read Hero'}}}}
FAILED tests/test_tutorial/test_fastapi/test_relationships/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...or'}}, ...}}}} == {'components'...or'}}, ...}}}}
FAILED tests/test_tutorial/test_fastapi/test_response_model/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...eate Hero'}}}} == {'components'...eate Hero'}}}}
FAILED tests/test_tutorial/test_fastapi/test_session_with_dependency/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...or'}}, ...}}}} == {'components'...or'}}, ...}}}}
FAILED tests/test_tutorial/test_fastapi/test_simple_hero_api/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...eate Hero'}}}} == {'components'...eate Hero'}}}}
FAILED tests/test_tutorial/test_fastapi/test_teams/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...or'}}, ...}}}} == {'components'...or'}}, ...}}}}
FAILED tests/test_tutorial/test_fastapi/test_update/test_tutorial001.py::test_tutorial - AssertionError: assert {'components'...or'}}, ...}}}} == {'components'...or'}}, ...}}}}
===================================================================================================================================================================================
```

**Comment by tiangolo** [comment]
TL;DR: Good news, this is available in the just released SQLModel 0.0.12! 🎉 

---

Awesome, thanks a lot for all the work @farahats9! 🚀 

And thanks everyone for the comments, reviews, etc. ☕ 

@farahats9 I continued the work here with several things on top on PR https://github.com/tiangolo/sqlmodel/pull/700. I took your commits so you're co-author of that PR (and contributor to SQLModel, check your badge here next to your username, or in https://github.com/tiangolo/sqlmodel/graphs/contributors 😎 ).

I did a few extra things:

* I reverted a few things, e.g. altering the docs to workaround for a previous lack of support for forward references.
* Added support/fix for forward reference errors.
* Removed several sections/modules that are no longer necessary, as were mainly to add type annotations to SQLAlchemy things, but now with SQLAlchemy's new proper types that's no longer necessary and I can reuse SQLAlchemy's things directly.
* Updated all needed tests, e.g. to support the latest FastAPI.
* Updated all the generated selects to use the new SQLAlchemy types.
* Updated the classes and type annotations with the new SQLAlchemy types, with some overloads to make sure they are compatible with SQLModel. E.g. `select().where()`.
* Redefine  all functions that would take column operations to accept the types used by SQLModel, e.g. `or_(Hero.name == "Deadpond")` as SQLAlchemy's version would require `Hero.name` to be a `Mapped[str]` instead of a `str`.
* I explored a lot if it was possible/sensible to use and encourage the new `Mapped[X]`, but sadly that breaks type annotations for creating new instances and dataclass_transform, as then that would expect to receive values with `Mapped[X]`, e.g. `Mapped[str]` instead of `str`. And it would also add complications to the Pydantic handling. So, this was a lot of work mostly exploratory, but not much ended in the PR.

It is now available in SQLModel 0.0.12. 🎉 

As this is now included in the repo and released, I'll now close this PR. Thanks! 🍰

**Comment by gdoumenc** [comment]
I had the same issue because the relationsship was always defined and I didn' set it Optional :
